src/ORM/LazyEagerLoader.php

```diff
@@ -0,0 +1,163 @@
+     * @return Cake\Datasource\EntityInterface|array
+     */
+    public function loadInto($entities, array $contain, Table $source)
+    {
+        $returnSingle = false;
+
+        if ($entities instanceof EntityInterface) {
+            $entities = [$entities];
+            $returnSingle = true;
+        }
+
+        $entities = new Collection($entities);
+        $query = $this->_getQuery($entities, $contain, $source);
+        $associations = array_keys($query->contain());
+
+        $entities = $this->_injectResults($entities, $query, $associations, $source);
+        return $returnSingle ? array_shift($entities) : $entities;
+    }
+
+    /**
+     * Builds a query for loading the passed list of entity objects along with the
+     * associations specified in $contain.
+     *
+     * @param Cake\Collection\CollectionInterface $objects
+     * @param array $contain The associations to be loaded
+     * @param Cake\ORM\Table $source The table to use for fetching the top level entities
+     * @return Cake\ORM\Query
+     */
+    protected function _getQuery($objects, $contain, $source)
+    {
+        $primaryKey = $source->primaryKey();
+        $method = is_string($primaryKey) ? 'get' : 'extract';
+
+        $keys = $objects->map(function ($entity) use ($primaryKey, $method) {
+            return $entity->{$method}($primaryKey);
+        });
+
+        $query = $source
+            ->find()
+            ->select((array)$primaryKey)
+            ->where(function ($exp, $q) use ($primaryKey, $keys, $source) {
+                if (is_array($primaryKey) && count($primaryKey) === 1) {
+                    $primaryKey = current($primaryKey);
+                }
+
+                if (is_string($primaryKey)) {
+                    return $exp->in($source->aliasField($primaryKey), $keys->toList());
+                }
+
+                $types = array_intersect_key($q->defaultTypes(), array_flip($primaryKey));
+                $primaryKey = array_map([$source, 'aliasField'], $primaryKey);
+                return new TupleComparison($primaryKey, $keys->toList(), $types, 'IN');
+            })
+            ->contain($contain);
+
+
+        foreach ($query->eagerLoader()->attachableAssociations($source) as $loadable) {
+            $config = $loadable->config();
+            $config['includeFields'] = true;
+            $loadable->config($config);
+        }
+
+        return $query;
+    }
+
+    /**
+     * Returns a map of property names where the association results should be injected
+     * in the top level entities.
+     *
+     * @param Cake\ORM\Table $source The table having the top level associations
+     * @param array $associations The name of the top level associations
+     * @return array
+     */
+    protected function _getPropertyMap($source, $associations)
+    {
+        $map = [];
+        foreach ($associations as $assoc) {
+            $map[$assoc] = $source->associations()->get($assoc)->property();
```

Comment:
You could save some function calls by having a local variable for the association collection.